cansav09/intro to R (#21)
AlexsLemonade/training-modules: intro_to_R_tidyverse/intro_to_tidyverse.Rmd

---
title: "Intro to Tidyverse"
output:   
  html_notebook: 
    toc: true
    toc_float: true
---

**CCDL 2019**

This Rmd is taking you through the basics of the tidyverse: a collection of packages
that make R use easier and more convenient.  

There are a lot of tutorials and free books Hadley Wickham and friends have on 
how to use the [tidyverse packages](https://r4ds.had.co.nz/index.html). 
[Datacamp](https://www.datacamp.com/) also has a bunch of good tutorials on this. 

## Set Up
R packages are collections of functions. But your R as it is will not have these
functions available unless you tell it to.  
In our docker container, we already have tidyverse installed for you. 
But if you needed to install it or other pacakges available on CRAN, you 
do it using the `install.packages` function like this: `install.packages("tidyverse")`
```{r Load tidyverse}
library(tidyverse)
```

> "GSE30074.tsv" is a file we will use for practicing the tidyverse.

Reviewer: Links to accession page on refine.bio

This data is from `refine.bio` and is gene expression data from medulloblastoma.
  
First let's read in the dataset using the tidyverse package: `readr`.
Although base R has read-in functions, the `readr` functions are faster and more
straight forward to use so we are going to use those here. 

> Because the file we are reading in is a .tsv file we will be using the 

Reviewer: You have .tsv here and `csv` below, should be consistent

Author: ? Where are you talking about?

Reviewer: Line 36

`read_tsv` function. 
But note that there are analogous functions for `csv` files and other files types.
```{r}

> # The file we are reading in, relative to our current directory, is in the "data"

Reviewer 2: Is talking about absolute and relative file paths too much at this point? Should that be addressed at some point? (Feels like a potential bug trap, if people aren't aware of absolute and relative paths)

# folder, so we need to put that as the argument. 

> gene.data <- read_tsv("data/GSE30074.tsv")

Reviewer: `file.path` ❗️

Reviewer 2: ![screen shot 2019-02-01 at 1 36 23 pm](https://user-images.githubusercontent.com/15315514/52142444-a9bf2580-2626-11e9-8705-ad0eb12110b3.png)

Is this expected?

Author: Yes

# Let's also read in the metadata for this gene expression data
metadata <- read_tsv("data/metadata_clean.tsv")
```

## dplyr pipes
A nifty thing with the tidyverse is the pipes: `%>%`

> This handy thing allows you to funnel the result of one expression to the next

Reviewer: Can we add a simpler example first? Prior to example 1 `without pipes`. Such as:
```R
filter(mtcars, cyl == 6)
mtcars %>% filter(cyl == 6)
```

making your code a little more streamlined.
This makes it so your code is cleaner and easier to read. 

> Let's look at an example with the built in data of `mtcars` of how the same 

Reviewer: Not sure why we couldn't start with the medulloblastoma metadata?

It's annoying and makes it harder for people to read. 

*Example 2:* Same result as 1 but with pipes!
```{r}
# Example of the same modifications as above but with pipes!
special.mtcars <- mtcars %>% 
                  rownames_to_column("car_model") %>% 
                  arrange(mpg) %>%
                  filter(mpg > 18) %>% 
                  select("mpg", "cyl", "gear")
```

Now that hopefully you are convinced that tidyverse helps your code to be neater and
easier to use and read, let's go through some of the popular tidyverse functions
and so we can create pipelines like this. 

## Common tidyverse functions

Let's say we wanted to filter this gene expression dataset to particular sample
groups.
In order to do this, we would use the function `filter` as well as a logic statement.
```{r}
# Here let's filter the data to have only females
metadata %>% filter(sex == "F")
```

We can use filter similarly for numeric statements.  
```{r}
# Here let's filter the data to have people only older than 
metadata %>% filter(age_mths > 100)
```

What if we wanted filter this dataset to males who are younger than 100 months? 
Let's make a longer pipe.
```{r }

> metadata %>% filter(age_mths > 100) %>%

Reviewer: ```suggestion
metadata %>% filter(age_mths > 100, sex == "M") %>%
```

youngest to oldest.
In order to do this, we would use the function `arrange` as well as the variable
we would like to sort by (in this case `age_mths`).
```{r}
metadata %>% arrange(age_mths) 
```

What if we want the the oldest patients at the top of this dataset instead? 
We can use the same function, but instead use the `desc` function too. 
```{r}
metadata %>% arrange(desc(age_mths))
```

What if we want to obtain summary statistics based on groups? 
Conveniently named, there's a function called `group_by` that seamlessly allows
us to do this. 
Also note that you can group by multiple variables at a time if you want to.
```{r}
metadata %>% group_by(sex) %>% 
             summarize(mean(age_mths))
```
What if we would like to create new variables based on other variables? 
For that we use `mutate` function. 
Let's convert the ages that are in months to years and make that into a new variable.
Here we are also using the `round` function from base R to limit the number of 
decimal points in our new variable to 2.
```{r}
metadata %>% mutate("age_yrs" = round(age_mths/12, 2)) %>%
             select(sample_id, age_mths, age_yrs)
```

> ## Gene data examples

Reviewer: I don't know that this section really adds too much over the metadata section (that we don't cover elsewhere) and has potential to be confusing.

Reviewer 2: I think it might be useful to see how one could also work with gene expression data.

## Set Up
R packages are collections of functions. 
Installing R on your computer installs `base` functions like statistical tests 
and some plotting functionality.
The `tidyverse` is a collection of packages that are handy for general data 
wrangling, analysis, and visualization. 
Other packages that are specifically handy for different biological analyses are 
found on [Bioconductor](https://www.bioconductor.org/).
If we want to use a package's functions we first need to install them.
In our Docker container, we already have `tidyverse` and other packages we will
use in this workshop installed for you. 
But if you needed to install it or other packages available on CRAN, you 
do it using the `install.packages` function like this: `install.packages("tidyverse")`
```{r Load tidyverse}
library(tidyverse)
```

"metadata.tsv" is a file we will use for practicing the tidyverse.
This metadata is from GSE30074 [refine.bio](https://www.refine.bio), a gene 
expression dataset from medulloblastoma.
  
First let's read in the dataset using the tidyverse package: `readr`.
Although base R has read-in functions, the `readr` functions are faster and more

> straight forward to use so we are going to use those here. 

Reviewer: ```suggestion
straightforward to use so we are going to use those here. 
```

Because the file we are reading in is a .tsv file we will be using the 
`read_tsv` function. 
But note that there are analogous functions for .csv files (`read_csv`) and 
other files types.

### About directories
Before we can import the file we need though, we should double check what
directory R is in, aka what our current *directory* is. 
We can do this by using the `getwd()` function, which will tell us what folder
we are in. 
Then we can use the `dir()` function to have R list the files in our current 
directory. 
What we should see is the file `metadata.tsv` show up as one of the files when 
we run `dir()`.
If for some reason we were not in the correct directory, we can either set our
directory to somewhere else using `setwd()` OR we can tell R to look for the 
file `metadata.tsv` somewhere else using a relative `file.path()`.

> We will discuss and practice these concepts more in other modules as we go along.

Reviewer: I don't believe we explicitly talk about relative vs. absolute paths anywhere else (and even if we do, repetition is helpful), so I think this should be a bit more in depth here.

Reviewer: Alternatively, include as part of intro (#25)

Author: I think this section is a good a time as any for them to practice it, but we might wanna introduce the concept in the slides (#25).

Author: I'll add more details on this section though too.

Author: Per our in person convoy: I'm gonna hold off on expanding this and other sections much more before we get what we have right now pushed.

```{r}
# Let's check what directory we are in:
getwd()

# Let's check what files are here
dir()

# The file we are reading in is our argument 
metadata <- read_tsv("metadata.tsv")
```

## dplyr pipes
A nifty thing with the tidyverse is the pipes: `%>%`
This handy thing allows you to funnel the result of one expression to the next
making your code a little more streamlined.

For example, the output from this:  
```
filter(metadata, sex == "F")
```  
...is the same as the output from this:  
```
metadata %>% filter(sex == "F")
```  
  
This makes it so your code is cleaner and easier to read. 
Let's look at an example with our metadata of of how the same 
functions look with or without pipes:

*Example 1:* without pipes:   
```{r}
metadata_edit <- metadata
metadata_edit <- arrange(metadata_edit, histology)
metadata_edit <- filter(metadata_edit, follow_up_mths > 50)
metadata_edit <- select(metadata_edit, sample_id, sex, organism)
```

> UGH, we have to repeat and assign `special.mtcars` so many times here! 

Reviewer: ```suggestion
UGH, we have to repeat and assign `metadata_edit` so many times here! 
```

It's annoying and makes it harder for people to read. 

*Example 2:* Same result as 1 but with pipes!
```{r}
# Example of the same modifications as above but with pipes!
metadata_edit <- metadata %>%
                 arrange(histology) %>%
                 filter(follow_up_mths > 50) %>%

>                  select(sample_id, sex, organism)

Reviewer: Add a block where we're viewing `metadata_edit`?

Reviewer: Perhaps in both spots (here and at ~line 95). Might be nice to do something with `all.equal` as well.

Author: Cool. Will do.

```

Now that hopefully you are convinced that tidyverse helps your code to be neater and
easier to use and read, let's go through some of the popular tidyverse functions
and so we can create pipelines like this. 

## Common tidyverse functions

Let's say we wanted to filter this gene expression dataset to particular sample
groups.
In order to do this, we would use the function `filter` as well as a logic statement.
```{r}
# Here let's filter the data to have only females
metadata %>% filter(sex == "F")
```

We can use filter similarly for numeric statements.  
```{r}
# Here let's filter the data to have people only older than 
metadata %>% filter(age_mths > 100)
```

> What if we wanted filter this dataset to males who are younger than 100 months? 

Reviewer: I think this is filtering to _older_ than 100 months

Author: Whoops

> Let's make a longer pipe.

Reviewer: ```suggestion
Let's add more arguments to `filter`.
```

```{r }
metadata %>% filter(age_mths > 100, sex == "M")
```

Let's build on to this pipe and keep only some of these variables with the 
`select` function. 
```{r}

> metadata %>% filter(age_mths > 100) %>%

Reviewer: ```suggestion
metadata %>% filter(age_mths > 100, sex == "M") %>%
```
to match above

             filter(sex == "M") %>% 
             select(sample_id, histology, metastasis)
```

Let's say we wanted to arrange this dataset so that the patients were arranged
youngest to oldest.
In order to do this, we would use the function `arrange` as well as the variable
we would like to sort by (in this case `age_mths`).
```{r}
metadata %>% arrange(age_mths) 
```

What if we want the the oldest patients at the top of this dataset instead? 
We can use the same function, but instead use the `desc` function too. 
```{r}
metadata %>% arrange(desc(age_mths))
```

What if we want to obtain summary statistics based on groups? 
Conveniently named, there's a function called `group_by` that seamlessly allows
us to do this. 

> Also note that you can group by multiple variables at a time if you want to.

Reviewer: ```suggestion
Also note that `group_by` allows us to group by multiple variables at a time.
```

```{r}
metadata %>% group_by(sex) %>% 
             summarize(mean(age_mths))
```
What if we would like to create new variables based on other variables? 
For that we use `mutate` function. 
Let's convert the ages that are in months to years and make that into a new variable.
Here we are also using the `round` function from base R to limit the number of 
decimal points in our new variable to 2.
```{r}
metadata_edit <- metadata %>% 

>              mutate("age_yrs" = round(age_mths/12, 2)) %>%

Reviewer: Indentation is off here and the line below.

>              mutate("age_yrs" = round(age_mths/12, 2)) %>%

Reviewer: Don't know if you need `age_yrs` to be in quotation marks?

Author: You don't, but sometimes I like to put the quotes so it sticks out a little more that we are naming something. I'll get rid of the quotes though.
